documentation/internal_architecture.md

```diff
@@ -0,0 +1,129 @@
+
+### Target updates and overall architecture
+
+In the same way that the scrape discovery manager runs one discovery mechanism for each `scrape_config`, the scrape manager runs a corresponding scrape pool for each such section. Both identify these sections (e.g. across reloads and across the two components) via the section's `job_name` field, which has to be unique in a given configuration file. The discovery manager sends target updates for each `scrape_config` over a synchronization channel to the scrape manager, [which then applies those updates to the corresponding scrape pools](https://github.com/prometheus/prometheus/blob/v2.3.1/scrape/manager.go#L150-L173). Each scrape pool in turn runs one scrape loop for each target. The overall hierarchy looks like this:
+
+* Scrape manager
+  * Scrape pool for `scrape_config` 1
+    * Scrape loop for target 1
+    * Scrape loop for target 2
+    * Scrape loop for target 3
+    * [...]
+    * Scrape loop for target n
+  * Scrape pool for `scrape_config` 2
+    * [...]
+  * Scrape pool for `scrape_config` 3
+    * [...]
+  * [...]
+  * Scrape pool for `scrape_config` n
+    * [...]
+
+### Target labels and target relabeling
+
+Whenever the scrape manager receives an updated list of targets for a given scrape pool from the discovery manager, the scrape pool applies default target labels (such as `job` and `instance`) to each target and applies [target relabeling configurations](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#%3Crelabel_config%3E) to produce the final list of targets to be scraped.
+
+### Target hashing and scrape timing
+
+To spread out scrapes within a scrape pool and in a consistenly slotted way across the `scrape_config`'s scrape interval, each target is [hashed by its label set and its final scrape URL](https://github.com/prometheus/prometheus/blob/v2.3.1/scrape/target.go#L75-L82). This hash is then used to [choose a deterministic offset](https://github.com/prometheus/prometheus/blob/v2.3.1/scrape/target.go#L84-L98) within that interval.
+
+### Target scrapes
+
+Finally, a scrape loop periodically scrapes its targets over HTTP and tries to decode the received HTTP responses according to the [Prometheus text-based metrics exposition format](https://prometheus.io/docs/instrumenting/exposition_formats/). It then applies [metric relabeling configurations](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#%3Cmetric_relabel_configs%3E) to each individual sample and sends the resulting samples to the storage subsystem. Additionally, it tracks and stores the staleness of time series over multiple scrape runs, records [scrape health information](https://prometheus.io/docs/concepts/jobs_instances/#automatically-generated-labels-and-time-series) (such as the `up` and `scrape_duration_seconds` metrics), and performs other housekeeping tasks to optimize the appending of time series to the storage engine.
```

Comment:
I'd mention timeouts here, a scrape shouldn't take more than the interval. We get that question once every month or two.